### PR TITLE
Extensible classes support

### PR DESCRIPTION
**Pull request by kunitoki**
This PR aims at allowing C++ binded classes to be extensible from lua:

```cpp

    luabridge::getGlobalNamespace(L)
        .beginClass<X>("X", luabridge::extensibleClass)
            .addConstructor<void(*)()>()
        .endClass();

    int result = runLua(R"(
        function X:test()
            return 1
        end

        local x = X()
        result = x:test()
    )");

    assert(result == 1);
```

### CODE REVIEW

File: Source/LuaBridge/detail/CFunctions.h [hunk 1]
```diff
@@ -356,6 +356,41 @@ inline int read_only_error(lua_State* L)
...
+//=================================================================================================
+/**
+ * @brief
+ */
+inline int index_extended_class(lua_State* L)
+{
+    LUABRIDGE_ASSERT(lua_istable(L, lua_upvalueindex(1)));
+
+    if (! lua_isstring(L, -1))
+        luaL_error(L, "%s", "invalid extensible class");
```
Comment (kunitoki) [inline Source/LuaBridge/detail/CFunctions.h:368]: TODO: improve the error message

File: Source/LuaBridge/detail/CFunctions.h [hunk 2]
```diff
@@ -356,6 +356,41 @@ inline int read_only_error(lua_State* L)
...
+    lua_pushvalue(L, lua_upvalueindex(1));
+    rawgetfield(L, -1, key);
+
+    return 1;
+}
+
+inline int newindex_extended_class(lua_State* L)
+{
+    if (! lua_istable(L, -3) || ! lua_isstring(L, -2))
+        luaL_error(L, "%s", "invalid extensible class");
```
Comment (kunitoki) [inline Source/LuaBridge/detail/CFunctions.h:382]: TODO: improve the error message

File: Source/LuaBridge/detail/FlagSet.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
+// https://github.com/kunitoki/LuaBridge3
```
Comment (kunitoki) [inline Source/LuaBridge/detail/FlagSet.h:1]: TODO: unit tests missing

**Comment by rpatters1** [comment]
If a base class sets this option, will all subclasses be extensible as well? (Just trying to understand what is being added.)

**Comment by kunitoki** [comment]
This is something i'm working on. Extensibility is a class property, so should be enabled in each class individually.

**Comment by rpatters1** [comment]
> Extensibility is a class property, so should be enabled in each class individually.

Yes, I agree. I suppose the better, more precise question is, will subclasses be able to access methods added to one of its base classes (up the hierarchy)?

**Comment by kunitoki** [comment]
Take a look at the tests https://github.com/kunitoki/LuaBridge3/pull/116/files#diff-cd70549e26a819fe4ab5502e1e783320b16be7b19c74582267b8cc90b73943e7R2684

File: Tests/Source/ClassTests.cpp [hunk 4]
```diff
@@ -2678,6 +2681,140 @@ TEST_F(ClassTests, NewIndexFallbackMetaMethodFreeFunctor)
...
+{
+    luabridge::getGlobalNamespace(L)
+        .beginClass<ExtensibleBase>("ExtensibleBase", luabridge::extensibleClass)
+            .addConstructor<void(*)()>()
+            .addFunction("baseClass", &ExtensibleBase::baseClass)
+        .endClass()
+    ;
+
+    runLua(R"(
+        function ExtensibleBase:baseClass() return 42 end
```
Comment (kunitoki) [inline Tests/Source/ClassTests.cpp:2810]: TODO: This should probably not be allowed, by checking if the method already exist in the class table.

**Comment by rpatters1** [comment]
I'm assuming these function all receive a `self` parameter. It might be good to include some tests around that.

**Comment by kunitoki** [comment]
> I'm assuming these functions all receive a self parameter. It might be good to include some tests around that.

Definately, this is just the beginning, a lot more tests are needed down there.

**Comment by rpatters1** [comment]
Another point of clarification. It looks like the answer is "no" but just to confirm:

Does this allow you to replace existing functions that were registered on the class? One of the reasons we added the `mixin` library was to improve the DX around clunky APIs in the original C++ classes. An example is that the C++ framework has its own homegrown string class, and in order to pass text to any of those classes, you have to stop and create one of the proprietary string types, stuff your text into it, then pass it. The mixin library replaces those methods with methods that take anything it can recognize as a string.

**Comment by kunitoki** [comment]
I'm thinking this by default should not be allowed (added a check for this here https://github.com/kunitoki/LuaBridge3/pull/116/commits/810813ebdb53786d62a4549ca511d68fca12cc69#diff-a6c1434d943944028ce926611075458078b52d6d2de84188f557c344c3c0fe74R297-R313 which prevents any modification).

But i see the use cases in controlled environments (we might have a use for that as well for the reason you mention). One complication for that feature could be allowing still to access the `super()` method being wrapper, in order to properly "proxy" it.

**Comment by ThistleSifter** [comment]
Will this also eventually include the ability to create custom subclasses from extensible classes?

**Comment by kunitoki** [comment]
Not sure i would mix luabridge inheritance (that works on top of c++ userdata) with mixed one (a lua table which parent is a c++ userdata), mostly because it will overcomplicate key resolution. But contributions are welcome. I suggest we start a discussion on how wewould like this to work

**Comment by kunitoki** [comment]
> Does this allow you to replace existing functions that were registered on the class?

This is now in https://github.com/kunitoki/LuaBridge3/blob/49ddf4f0ef3cc1210b38de75c56ac53f019b1a64/Tests/Source/ClassTests.cpp#L2870-L2895

I would maybe later enable separate index and newindex metamethods for extensible classes, so to not incur into a performance penalty for classes that are not extensible.

**Pull request merged**